src/sa.py

```diff
@@ -1,15 +1,94 @@
 import argparse
+from fasta_dict import fasta_func
+from fastq_dict import fastq_func
+from tree import construct_sorted_tree, subtree_labels
 
+def construct_array(x):
+    '''Return the suffix array of string x
+    
+    >>> construct_array('mississippi')
+    [11, 10, 7, 4, 1, 0, 9, 8, 6, 3, 5, 2]
+    '''
+    T = construct_sorted_tree(x)
+    L = subtree_labels(T[0])
+
+    return list(L)
+
+def recursive_lower_bound(p, x, sa, start, end, i_letter):
+    '''Function recursive_lower_bound that takes a pattern p, string x, suffix 
+    array sa of x, to recursively find a the lower bound of p[:i_letter] in sa'''
+
+    if start == end:
+        return start
+    else:
+        mid = (end+start)//2
+        if p[i_letter] <= x[sa[mid]+i_letter]:
+            end = mid
+        else:
+            start = mid+1
+    
+    return recursive_lower_bound(p, x, sa, start, end, i_letter)
+
+def recursive_upper_bound(p, x, sa, start, end, i_letter):
+    '''Function recursive_lower_bound that takes a pattern p, string x, suffix 
+    array sa of x, to recursively find a the upper bound of p[:i_letter] in sa'''
+
+    if start == end:
+        return start
+    else:
+        mid = (end+start)//2
+        if p[i_letter] >= x[sa[mid]+i_letter]:
+            start = mid+1
+        else:
+            end = mid
+    
+    return recursive_upper_bound(p, x, sa, start, end, i_letter)
```

Comment:
When you are searching for a single letter, `a`, you have `upper_bound(x, a) == lower_bound(x, a+1)` so you could have saved a little work here.